lib/models/LifecycleConfiguration.js

```diff
@@ -0,0 +1,653 @@
+const assert = require('assert');
+const UUID = require('uuid');
+
+const errors = require('../errors');
+
+/**
+ * Format of xml request:
+
+ <LifecycleConfiguration>
+    <Rule>
+        <ID>id1</ID>
+        <Filter>
+            <Prefix>logs/</Prefix>
+        </Filter>
+        <Status>Enabled</Status>
+        <Expiration>
+            <Days>365</Days>
+        </Expiration>
+    </Rule>
+    <Rule>
+        <ID>DeleteAfterBecomingNonCurrent</ID>
+        <Filter>
+            <And>
+                <Prefix>logs/</Prefix>
+                <Tag>
+                    <Key>key1</Key>
+                    <Value>value1</Value>
+                </Tag>
+            </And>
+        </Filter>
+        <Status>Enabled</Status>
+        <NoncurrentVersionExpiration>
+            <NoncurrentDays>1</NoncurrentDays>
+        </NoncurrentVersionExpiration>
+        <AbortIncompleteMultipartUploads>
+            <DaysAfterInitiation>1</DaysAfterInitiation>
+        </AbortIncompleteMultipartUploads>
+    </Rule>
+</LifecycleConfiguration>
+ */
+
+/**
+ * Format of config:
+
+  config = {
+    rules = [
+        {
+            ruleID: <value>,
+            ruleStatus: <value>,
+            filter: {
+                rulePrefix: <value>,
+                tags: [
+                    {
+                        key: <value>,
+                        val: <value>
+                    },
+                    {
+                        key: <value>,
+                        val: <value>
+                    }
+                ]
+            },
+            actions: [
+                {
+                    actionName: <value>,
+                    days: <value>,
+                    date: <value>,
+                    deleteMarker: <value>
+                },
+                {
+                    actionName: <value>,
+                    days: <value>,
+                    date: <value>,
+                    deleteMarker: <value>,
+                },
+            ]
+        }
+    ]
+  };
+ */
+
+class LifecycleConfiguration {
+    /**
+     * Create a Lifecycle Configuration instance
+     * @param {string} xml - the parsed xml
+     * @return {object} - LifecycleConfiguration instance
+     */
+    constructor(xml) {
+        this._parsedXML = xml;
+        this._ruleIDs = [];
+        this._tagKeys = [];
+        this._config = {};
+    }
+
+    /**
+     * Get the lifecycle configuration
+     * @return {object} - the lifecycle configuration
+     */
+    getLifecycleConfiguration() {
+        const rules = this._buildRulesArray();
+        if (rules.error) {
+            this._config.error = rules.error;
+        }
+        return this._config;
+    }
+
+    /**
+     * Build the this._config.rules array
+     * @return {object} - contains error if any rule returned an error
+     * or parsing failed
+     */
+    _buildRulesArray() {
+        const rules = {};
+        this._config.rules = [];
+        if (!this._parsedXML || this._parsedXML === '') {
```

Comment:
you can go with the first test only, as an empty string is true-ish (`!''` is true), though considering that you access attributes of `_parsedXML` next, maybe a better validation can be to check that it is a JS object (`typeof this._parsedXML === 'object'`) otherwise it would raise an exception when getting the `LifecycleConfiguration` attribute if the user passes a non-empty string.

As a general comment for this whole function, using the `joi` module to do the first pass of validation with declarative syntax, then doing the few domain-specific checks not supported by joi would likely lead to much more compact and maintainable code (if you decide to go this way, have a look at how it's used in backbeat + reference at https://www.npmjs.com/package/joi, otherwise let me know I will happily review the remaining parsing code, I'll skip it for now as it's a significant chunk to review that will be wasted time if you choose to rewrite with `joi` :wink: )

Reply:
Yeah, I saw that module in backbeat, but I thought it would be better not to use it since it's not used anywhere else in the Arsenal repo. I can rewrite it if you think it's worth it though :)

Reply:
I personally think it's worth it, because it can both simplify the code a lot and get rid of potential issues in the current code (because of its length and relative complexity) and I think it was on Rahul's mind concerning the refactoring projects (using joi where configuration parsing occurs).

Now, I find it rather simple and pleasant to use joi module but there's some small effort anyway to get into it, and it's always a bit annoying to rewrite a piece of code we just wrote, so feel free to do the change if you have some time for it but I'm not pushing either, as long as we have working code :)